KAFKA-17813: Moving broker endpoint class and common server connection id

apoorvmittal10 opened the pull request:
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apoorvmittal10 committed:
KAFKA-17813: Moving broker endpoint class and common server connection id

chia7712 commented:
@apoorvmittal10 I love this cleanup and migration. Could you please rebase code to fix conflicts as #17473 is already merged

apoorvmittal10 commented:
> @apoorvmittal10 I love this cleanup and migration. Could you please rebase code to fix conflicts as #17473 is already merged

Thanks @chia7712, sure. I ll force push the branch once https://github.com/apache/kafka/pull/17474 is also merged, as the changes are on top of that. Hence keeping the PR in draft mode till then.

apoorvmittal10 commented:
@chia7712 Yeah, sorry https://github.com/apache/kafka/pull/17474 is not required for this PR. I have rebased it and marked for review. Please if you could take a look.

apoorvmittal10 commented:
Looking into the tests failure.

brandboat commented on clients/src/test/java/org/apache/kafka/common/network/BrokerEndPointTest.java:
```diff
@@ -0,0 +1,57 @@
+        assertEquals("myhost", brokerEndPoint.get().host());
+        assertEquals(9092, brokerEndPoint.get().port());
+
+        brokerEndPoint = BrokerEndPoint.parseHostPort("127.0.0.1:9092");
```
It would be great if we can add one more `[ipv6_host]:port` test case.

apoorvmittal10 replied:
Added a test case.

brandboat commented on clients/src/main/java/org/apache/kafka/common/network/BrokerEndPoint.java:
```diff
@@ -0,0 +1,98 @@
+ */
+public class BrokerEndPoint {
+
+    private static final Pattern URI_PARSE_EXP = Pattern.compile("\\[?([0-9a-zA-Z\\-%._:]*)]?:([0-9]+)");
```
This seems slightly different from the original Scala version, perhaps this should be `\\[?([0-9a-zA-Z\\-%._:]*)\\]?:([0-9]+)` ?

apoorvmittal10 replied:
This is not needed, additional escape for `]` is not needed in java.

brandboat replied:
Thanks for explanation ! I also notice that we already have a BrokerEndPointTest.scala which cover some tests... Anyway, thanks for your patience.

apoorvmittal10 replied:
Yeah, it's a good test file (BrokerEndPointTest.scala). I thought to port that too in java but it needs some zk specific scala classes to parse and uses additional scala classes which I cannot reference.

brandboat reviewed (commented):
Thanks for the refactor !!!

chia7712 commented on clients/src/main/java/org/apache/kafka/common/network/ServerConnectionId.java:
```diff
@@ -0,0 +1,120 @@
+        }
+
+        try {
+            return BrokerEndPoint.parseHostPort(split[0])
```
```java
            return BrokerEndPoint.parseHostPort(split[0])
                .flatMap(localHost -> BrokerEndPoint.parseHostPort(split[1])
                    .map(remoteHost -> new ServerConnectionId(localHost, remoteHost, Integer.parseInt(split[2]), Integer.parseInt(split[3]))));
```
we don't need to wrap it by `Optional.of`

chia7712 replied:
Additionally, `BrokerEndPoint.parseHostPort` is used by `ServerConnectionId` only, so maybe we can inline it?

apoorvmittal10 replied:
Thanks for the suggestion, I have addressed the first comment. Sorry I didn't understand the `inline` comment.

junrao replied:
I think Chia is suggesting to move `BrokerEndPoint.parseHostPort` as a static method in this class. Since the connection string doesn't contain brokerId. It doesn't quite fit into `BrokerEndPoint`.

apoorvmittal10 replied:
I have moved it same class now.

chia7712 commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1221,7 +1206,7 @@ private[kafka] class Processor(
       try {
-        val remoteHost = ConnectionId.fromString(connectionId).getOrElse {
+        val remoteHost = ServerConnectionId.fromString(connectionId).orElseThrow { () =>
           throw new IllegalStateException(s"connectionId has unexpected format: $connectionId")
```
Maybe `ServerConnectionId.fromString` can throw `IllegalStateException` instead of returning `Optional`?

apoorvmittal10 replied:
I was trying to maintain least refactor while keeping original context. I ll look into this as well.

chia7712 reviewed (commented):
@apoorvmittal10 thanks for this refactor

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-17813

apoorvmittal10 commented:
@chia7712 @brandboat Thanks for the review, can you please re-look.

brandboat reviewed (approved):
LGTM, thanks !

junrao commented on clients/src/main/java/org/apache/kafka/common/network/ServerConnectionId.java:
```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.network;
```
This is only used on the server. Could we put it under the server-common module?

apoorvmittal10 replied:
Make sense, but then we have to remove the `generateConnectionId` method from this class as that's used by `NioEchoServer`. And if we remove that then there is no consolidation between Selector.java and SocketServer.scala related connectionId methods. Wdyt?

junrao replied:
Got it. We can keep it in the client module then.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/ServerConnectionId.java:
```diff
@@ -0,0 +1,120 @@
+/**
+ * ServerConnectionId is used to uniquely identify a connection between a client and a server.
+ */
+public class ServerConnectionId {
```
ServerConnectionId => ConnectionId ?

junrao replied:
If we keep this class in the client module, it's better to give it a more accurate name. Perhaps ConnectionIdOnServer?

apoorvmittal10 replied:
Hmmm, I somehow felt `ServerConnectionId` is better as it prefixes with `server` itself. I can write the class description appropraitely. If you still think it's not appropriate then I ll move it to `ConnectionIdOnServer`. Please let me know.

chia7712 replied:
`ServerConnectionId` needs to remain in the clients module since it's also used by the clients' test code. Perhaps we could rewrite `NioEchoServer` to use a different connection ID scheme to decouple it. In #16460, we wanted to add a more meaningful string, but that's not necessary for aligning with the core server code.

chia7712 replied:
@apoorvmittal10 what do you think abort moving `ServerConnectionId` to server module?

apoorvmittal10 replied:
@chia7712 That we can't do, as we discussed that here: https://github.com/apache/kafka/pull/17519#discussion_r1805238289

junrao commented on clients/src/main/java/org/apache/kafka/common/network/ServerConnectionId.java:
```diff
@@ -0,0 +1,120 @@
+        try {
+            return BrokerEndPoint.parseHostPort(split[0])
+                .flatMap(localHost -> BrokerEndPoint.parseHostPort(split[1])
+                    .map(remoteHost -> new ServerConnectionId(localHost, remoteHost, Integer.parseInt(split[2]), Integer.parseInt(split[3]))));
```
indentation

apoorvmittal10 replied:
I have changed the way it's written now.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/BrokerEndPoint.java:
```diff
@@ -0,0 +1,98 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.network;
```
This is only used on the server. Could we put it under the server-common module?

apoorvmittal10 replied:
As the other class is in `clients` uses BrokerEnpoint now henc kept it in clients itself, under network. As that's the only place I found to aggregate where clients and server related classes can go.

apoorvmittal10 replied:
Moved to server-common as we have now moved to map entry as per @chia7712's comment.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the PR. Left a few comments.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
I have addressed the comments, 1 with the name is pending. If the name doesn't seems right for `ServerConnectionId` then I ll change it. Though I have added the class description.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/network/ServerConnectionId.java:
```diff
@@ -0,0 +1,143 @@
+     * Note that unlike EndPoint (or listener) this URI has no security information.
+     */
+    // Visible for testing
+    static Optional<BrokerEndPoint> parseHostPort(String connectionString) {
```
Could you please replace `BrokerEndPoint` by `Map.Entry`? we are under java 11, so it is easier to create `Entry` now.

apoorvmittal10 replied:
Done, though not a big fan of Map.Entry but seems sensible here as there was no broker id for BrokerEndpoint.

apoorvmittal10 committed:
Moving to map entry

apoorvmittal10 commented:
I have resolved the comments, build passed.

junrao commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -25,7 +25,7 @@ import java.util
 import java.util.concurrent._
 import java.util.concurrent.atomic._
-import kafka.cluster.{BrokerEndPoint, EndPoint}
+import kafka.cluster.{EndPoint}
```
We can remove the brackets since there is a single import.

apoorvmittal10 replied:
Missed it, thought spotlessApply will fix. Done.

junrao commented on clients/src/test/java/org/apache/kafka/common/network/ServerConnectionIdTest.java:
```diff
@@ -110,4 +110,38 @@ public void testGenerateConnectionId() throws IOException {
+        assertEquals(9092, brokerEndPoint.get().port());
+
+        // IPv6 endpoint
+        brokerEndPoint = ServerConnectionId.parseHostPort("[2001:db8::1]:9092");
```
Hmm, an interesting thing is that if we take a ServerConnectionId with host name of [2001:db8::1] and call ServerConnectionId.fromString, we will get a slightly different ServerConnectionId since the hostname will not have the bracket. It's not an issue now, but it may cause confusion in the future.

chia7712 replied:
`ServerConnectionId` doesn't face ambiguity like a URL does, so in my opinion, the hostname in `ServerConnectionId` doesn't need brackets.

apoorvmittal10 replied:
Do I need to dix anything in this? I have just used the exisiting regex for parsing, from scala to java.

apoorvmittal10 replied:
Yeah, this is not required. I have pushed an additional test for same.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. A couple of more comments.

apoorvmittal10 committed:
Correcting imports

chia7712 reviewed (approved):
LGTM

junrao commented on clients/src/main/java/org/apache/kafka/common/network/ServerConnectionId.java:
```diff
@@ -0,0 +1,143 @@
+ */
+public class ServerConnectionId {
+
+    private static final Pattern URI_PARSE_EXP = Pattern.compile("\\[?([0-9a-zA-Z\\-%._:]*)]?:([0-9]+)");
```
A couple of things.
1. `Utils.HOST_PORT_PATTERN` is more general than this. We could just reuse the former instead.
2. It seems that we have an existing bug. `InetAddress.getHostAddress()` doesn't contain brackets. When we generate a parsable string for host:port, we need to add brackets for the host part since it can contain colon. For example, we need to do that for bootstrap.servers. However, in Processor, when generating a connection string, we just concatenate the host and the port without adding the brackets. This means that when we try to convert the connection string back to ConnectionId, we can't parse the hostname properly if it's IPV6. To fix this, it seems that we need to add brackets to the hostname when generating the connection string.

apoorvmittal10 replied:
> Utils.HOST_PORT_PATTERN is more general than this. We could just reuse the former instead.

I initially looked at it but didn't use that as it also has `protocol` parsing which we don't require. And moreover I thought to port same scala version to java.

> This means that when we try to convert the connection string back to ConnectionId, we can't parse the hostname properly if it's IPV6. To fix this, it seems that we need to add brackets to the hostname when generating the connection string.

I don't think it's a bug, I am writing a test case for same as well. The regex splits on the end `:` hence the host and port will be constructed correctly.

apoorvmittal10 replied:
I have pushed a test case as well for ipv6.

junrao replied:
Thanks for the explanation. It's fine to have a separate regex just for ServerConnectionId. If so, the host name in the connection string will never have brackets in it, right? Then, it would be useful to change the regex, the comment and the tests to remove the support on brackets accordingly.

apoorvmittal10 replied:
>If so, the host name in the connection string will never have brackets in it, right?

I am not sure about this but the code handles with or without brackets. I think it's the right behaviour. The test cases I wrote additionally verifies that.

apoorvmittal10 replied:
Also just trying to be consistent to the scala version of the previous code.

junrao replied:
> I am not sure about this but the code handles with or without brackets. I think it's the right behaviour. 

Well, we create the connection string in two paths. The first one gets the hostname from `InetAddress.getHostAddress()`, which shouldn't include brackets according to https://groups.google.com/g/golang-dev/c/AiC3MffM4-U. The second one gets the hostname from ServerConnectionId.parseHostPort() which stripes out the brackets if present.

apoorvmittal10 replied:
Sure @junrao, I have made the change.

apoorvmittal10 commented:
@junrao Thanks for the feedback and review. Please let me know if I need to address more. Build has passes for the PR.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. A couple of more comments.

apoorvmittal10 committed:
Adding ipv6 additional test

apoorvmittal10 committed:
Removing brackets from regex

apoorvmittal10 committed:
Correcting comment

apoorvmittal10 commented:
@junrao I have addressed the comment, removed brackets, corrected tests and comments. Please can you review.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/ServerConnectionId.java:
```diff
@@ -0,0 +1,146 @@
+    // The regex for parsing the URI string. The URI string should be in the format of "host:port",
+    //  where host can be an IPv4 address or an IPv6 address.
+    //  Note: The IPv6 address should not be enclosed in square brackets.
+    private static final Pattern URI_PARSE_EXP = Pattern.compile("([0-9a-zA-Z\\-%._:]*):([0-9]+)");
```
Since this is not really a URI, should we rename URI_PARSE_EXP to sth like HOST_PORT_PARSE_EXP?

apoorvmittal10 replied:
Done.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Just a minor comment.

apoorvmittal10 committed:
Addressing minor comments

junrao reviewed (approved):
@apoorvmittal10 : Thanks for the updated PR. LGTM
